events/events.go

```diff
@@ -0,0 +1,379 @@
+package events
+
+import (
+	json "encoding/json"
+	"errors"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	manifold "github.com/manifoldco/go-manifold"
+	merrors "github.com/manifoldco/go-manifold/errors"
+	"github.com/manifoldco/go-manifold/idtype"
+)
+
+// Type represents the different types of events.
+type Type string
+
+const (
+	// TypeUserCreated represents a user creation
+	TypeUserCreated Type = "user.created"
+
+	// TypeOperationProvisioned represents a provision operation
+	TypeOperationProvisioned Type = "operation.provisioned"
+
+	// TypeResourceCreated represents a resource creation
+	TypeResourceCreated Type = "resource.created"
+)
+
+// State represents the state of an event. Events usually starts only
+// containing raw references to the system and later are processed to expand
+// objects information.
+type State string
+
+const (
+	// StatePending represents the event information is pending expansion. That's
+	// done as a worker job. Users don't have access while an event is pending.
+	StatePending State = "pending"
+
+	// StateTracking represents the event is sending information to a 3rd-party
+	// analytics. Users can access an event while is tracking state.
+	StateTracking State = "tracking"
+
+	// StateDone represents the event information expansion is done. The event
+	// is immutable going forward. Users can access a done event.
+	StateDone State = "done"
+)
+
+// SourceType represents where the request came from.
+type SourceType string
+
+const (
+	// SourceDashboard is a request coming from the dashboard
+	SourceDashboard SourceType = "dashboard"
+
+	// SourceCLI is a request coming from the cli
+	SourceCLI SourceType = "cli"
+
+	// SourceSystem is an internal request
+	SourceSystem SourceType = "system"
```

Comment:
Should we consider breaking this out so we have a source for each of our micro-services.
Or an additional field where we can populate that info, and for CLI and Dashboard it would just repeat.
```
SourceType = "dashboard"
SourceService = "dashboard"
```
```
SourceType = "system"
SourceService = "identity"
```
Something like that?

Reply:
Do you think it would be valuable? I was thinking more from a user point of view. Ie: was this event created by an actor, or was it automated?
I don't think a system event will come from more than one place. So we could infer from the type the source.

Reply:
Yeah that's true :+1: good point.